pkg/commands/setnamespace.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/kubernetes-sigs/kustomize/pkg/constants"
+	"github.com/kubernetes-sigs/kustomize/pkg/fs"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+type setNamespaceOptions struct {
+	namespace string
+}
+
+// newCmdSetNamespace sets the value of the namespace field in the kustomization.
+func newCmdSetNamespace(fsys fs.FileSystem) *cobra.Command {
+	var o setNamespaceOptions
+
+	cmd := &cobra.Command{
+		Use:   "namespace",
+		Short: "Sets the value of the namespace field in the kustomization file",
+		Example: `
+The command
+	set namespace staging
+will add the field "namespace: staging" to the kustomization file if it doesn't exist,
+and overwrite the value with "staging" if the field does exist.
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := o.Validate(args)
+			if err != nil {
+				return err
+			}
+			return o.RunSetNamespace(fsys)
+		},
+	}
+	return cmd
+}
+
+// Validate validates setNamespace command.
+func (o *setNamespaceOptions) Validate(args []string) error {
+	if len(args) != 1 {
+		return errors.New("must specify exactly one namespace value")
+	}
+	ns := args[0]
+	if errs := validation.IsDNS1123Label(ns); len(errs) != 0 {
```

Comment:
Does namespace follow this label format? Is there any doc for that?

Reply:
I'm not aware of a doc but I used this validation function after a little experimentation and reading some source code:

```
$ kubectl create ns _test_    
The Namespace "_test_" is invalid: metadata.name: Invalid value: "_test_": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')
```

I traced that error through:

https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L108
https://github.com/kubernetes/apimachinery/blob/master/pkg/api/validation/generic.go#L62
https://github.com/kubernetes/kubernetes/blob/e38efdcce6860141609bb8b9f959d06d6643fdf5/pkg/apis/core/validation/validation.go#L4996

But let me know if I missed something.

Reply:
Good tracing and that gives enough info.